src/Microsoft.OData.Core/UriParser/Binders/FunctionCallBinder.cs

```diff
@@ -234,7 +234,27 @@ internal QueryNode BindFunctionCall(FunctionCallToken functionCallToken)
 
             // If there isn't, bind as Uri function
             // Bind all arguments
-            List<QueryNode> argumentNodes = new List<QueryNode>(functionCallToken.Arguments.Select(ar => this.bindMethod(ar)));
+            List<QueryNode> argumentNodes = functionCallToken.Arguments.Select(argument =>
+            {
+                // If the function is IsOf or Cast and the argument is a dotted identifier, we need to bind it differently
+                if (UnboundFunctionNames.Contains(functionCallToken.Name) && argument.ValueToken is DottedIdentifierToken dottedIdentifier)
+                {
+                    // Find the type of the dotted identifier by resolving it against the model. This also ensure case-insensitive resolution.
+                    IEdmSchemaType dottedIdentifierType = UriEdmHelpers.FindTypeFromModel(state.Model, dottedIdentifier.Identifier, this.Resolver);
+
+                    // If the dotted identifier is a primitive type and the next token is not null, we need to ensure next token is null
+                    if (dottedIdentifierType is IEdmPrimitiveType && dottedIdentifier.NextToken != null)
+                    {
+                        // Push the current argument onto the stack to keep track of the previous argument
+                        dottedIdentifier.NextToken = null;

```

Comment:
Is it safe to modify tokens like this? Is this something we do elsewhere in the code?

Reply:
Instead of `dottedIdentifier.NextToken = null` I have opted to do the following:
```cs
// If the next token is not null, we need to ensure it is null
dottedIdentifier = new DottedIdentifierToken(dottedIdentifier.Identifier, null);
```
This is done in [ParseIdentifier(QueryToken parent)](https://github.com/OData/odata.net/blob/main/src/Microsoft.OData.Core/UriParser/Parsers/IdentifierTokenizer.cs#L71) where `parent` is the `NextToken`  and for DottedIdentifier, parent is always null